graphbolt/include/graphbolt/cuda_ops.h

```diff
@@ -97,6 +98,51 @@ torch::Tensor ExclusiveCumSum(torch::Tensor input);
  */
 torch::Tensor UVAIndexSelectImpl(torch::Tensor input, torch::Tensor index);
 
+/**
+ * @brief Sample neighboring edges of the given nodes and return the induced
+ * subgraph.
+ *
+ * @param indptr Index pointer array of the CSC.
+ * @param indices Indices array of the CSC.
+ * @param nodes The nodes from which to sample neighbors.
+ * @param fanouts The number of edges to be sampled for each node with or
+ * without considering edge types.
+ *   - When the length is 1, it indicates that the fanout applies to all
+ * neighbors of the node as a collective, regardless of the edge type.
+ *   - Otherwise, the length should equal to the number of edge types, and
+ * each fanout value corresponds to a specific edge type of the node.
+ * The value of each fanout should be >= 0 or = -1.
+ *   - When the value is -1, all neighbors will be chosen for sampling. It is
+ * equivalent to selecting all neighbors with non-zero probability when the
+ * fanout is >= the number of neighbors (and replacement is set to false).
+ *   - When the value is a non-negative integer, it serves as a minimum
+ * threshold for selecting neighbors.
+ * @param replace Boolean indicating whether the sample is preformed with or
+ * without replacement. If True, a value can be selected multiple times.
+ * Otherwise, each value can be selected only once.
+ * @param layer Boolean indicating whether neighbors should be sampled in a
+ * layer sampling fashion. Uses the LABOR-0 algorithm to increase overlap of
+ * sampled edges, see arXiv:2210.13339.
+ * @param return_eids Boolean indicating whether edge IDs need to be returned,
+ * typically used when edge features are required.
+ * @param type_per_edge A tensor representing the type of each edge, if present.
+ * @param probs_or_mask An optional tensor with (unnormalized) probabilities
+ * corresponding to each neighboring edge of a node. It must be
+ * a 1D tensor, with the number of elements equaling the total number of edges.
+ * @param random_seed An optional random_seed parameter to fix the random_seed
+ * of the sampling operation. (Useful when layer=True)
+ *
+ * @return An intrusive pointer to a FusedSampledSubgraph object containing
+ * the sampled graph's information.
+ */
+c10::intrusive_ptr<sampling::FusedSampledSubgraph> SampleNeighbors(
```

Comment:
Can we make this an class method of fused_csc_sampling_graph?

Reply:
It is already being dispatched inside `FusedCSCSamplingGraph::SampleNeighbors`. This way, keeping different implementations in different files is easier, which helps compile times greatly. For example, this implementation takes a long time to compile due to the three nested type dispatches. Especially because the dispatch for `probs_or_mask` is done for all possible types (bool. int8, uint8, int16, int32, int64, float16, bfloat16, float32, float64, for a total of 10 types times 4 times 4 = 40 type combinations).

Reply:
Also, I might reuse this implementation for temporal sampling as well, in the FusedCSCSamplingGraph code, temporal sampling code path is separate.

Reply:
So TemporalSampleNeighbors might call this function as well.

Reply:
My concern is as an cuda_op in cuda_ops.h, it is overwhelming for the code to know the data type FusedSampledSubgraph, which belongs to fused_csc_sampling_graph.

Reply:
One solution is then to return a `std::tuple` and then form a fused_csc_sampling_graph inside the CPU source file.

Reply:
Encapsulating the implementation into a class will overall reduce readability I am afraid.

Reply:
Moreover, I have plans for later using this function when we have a graph cache.

Reply:
Or I can have `cuda_ops.h` as usual but create another header file called `cuda_sampling_ops.h`. Only this file can include the definition of the FusedSamplingGraph.

Reply:
Then cuda_ops.h does not have to import the contents of the FusedSampledSubgraph.

Reply:
made the changes.